cpp/vector_assembly.hpp

```diff
@@ -0,0 +1,121 @@
+  xtl::span<const std::uint32_t> cell_info;
+  if (needs_transformation_data)
+  {
+    mesh->topology_mutable().create_entity_permutations();
+    cell_info = xtl::span(mesh->topology().get_cell_permutation_info());
+  }
+
+  const std::vector<std::uint8_t>& perms = mesh->topology().get_facet_permutations();
+
+  // Assemble using dolfinx
+  dolfinx::fem::impl::assemble_exterior_facets<PetscScalar>(apply_dof_transformation, b, *mesh,
+                                                            active_facets, dofs, bs, kernel,
+                                                            constants, coeffs, cell_info, perms);
+}
+
+void assemble_cells(xtl::span<PetscScalar> b, std::shared_ptr<dolfinx::fem::FunctionSpace> V,
+                    const xtl::span<const std::int32_t>& active_cells, kernel_fn& kernel,
+                    const dolfinx::array2d<PetscScalar>& coeffs,
+                    const xtl::span<const PetscScalar>& constants)
+{
+  // Extract mesh
+  std::shared_ptr<const dolfinx::mesh::Mesh> mesh = V->mesh();
+
+  // Extract function space data
+  std::shared_ptr<const dolfinx::fem::DofMap> dofmap = V->dofmap();
+  const dolfinx::graph::AdjacencyList<std::int32_t>& dofs = dofmap->list();
+  const int bs = dofmap->bs();
+  std::shared_ptr<const dolfinx::fem::FiniteElement> element = V->element();
+  const std::function<void(const xtl::span<double>&, const xtl::span<const std::uint32_t>&,
+                           std::int32_t, int)>
+      apply_dof_transformation = element->get_dof_transformation_function<double>();
+
+  // NOTE: Need to reconsider this when we get to jump integrals between disconnected interfaces
+  const bool needs_transformation_data = element->needs_dof_transformations();
+
+  // Get permutation data
+  xtl::span<const std::uint32_t> cell_info;
+  if (needs_transformation_data)
+  {
+    mesh->topology_mutable().create_entity_permutations();
+    cell_info = xtl::span(mesh->topology().get_cell_permutation_info());
+  }
+
+  dolfinx::fem::impl::assemble_cells<PetscScalar>(apply_dof_transformation, b, mesh->geometry(),
+                                                  active_cells, dofs, bs, kernel, constants, coeffs,
+                                                  cell_info);
+}
+} // namespace
+
+namespace dolfinx_cuas
+{
+
+/// Assemble vector for given kernel function
+/// @param[in,out] b the vector to be assembled
+/// @param[in] V the function space
+/// @param[in] bcs Dirichlet boundary conditions
+/// @param[in] active_entities list of indices (local to process) of entities to be integrated over
+/// @param[in] kernel the custom integration kernel
+/// @param[in] coefficients used in the variational form
+/// @param[in] constants used in the variational form
+/// @param[in] type the integral type
+void assemble_vector(
+    xtl::span<PetscScalar> b, std::shared_ptr<dolfinx::fem::FunctionSpace> V,
+    const std::vector<std::shared_ptr<const dolfinx::fem::DirichletBC<PetscScalar>>>& bcs,
```

Comment:
As it seems like bcs are discared here, should we even send them in?

Reply:
Depends on whether one of the next steps is to make this work with bcs.

Reply:
I would suggest we follow the dolfin-x approach, and apply bcs through lifting, i.e.
https://github.com/FEniCS/dolfinx/blob/main/cpp/dolfinx/fem/assemble_vector_impl.h#L1018
https://github.com/FEniCS/dolfinx/blob/main/cpp/dolfinx/fem/assemble_vector_impl.h#L954
Then there are no bc input to the assemble_vector.